### ISSUE TITLE
[job-queue] Change type hint from Closure to callable

### ISSUE THREAD

**Issue by msheakoski** [php-enqueue/enqueue-dev#286]
In the JobRunner (and possibly other areas), the type hint currently limits you to working with closures. For more complex jobs, it is nicer to have an object because you can utilize DI and autoloading. Changing `Closure` to `callable` would maintain backwards compatibility and open up more options to the end user.

https://github.com/php-enqueue/enqueue-dev/blob/02f992c44871333bb31f85e93acd32b96425f093/pkg/job-queue/JobRunner.php#L34
https://github.com/php-enqueue/enqueue-dev/blob/02f992c44871333bb31f85e93acd32b96425f093/pkg/job-queue/JobRunner.php#L66
https://github.com/php-enqueue/enqueue-dev/blob/02f992c44871333bb31f85e93acd32b96425f093/pkg/job-queue/JobRunner.php#L81

**Comment by msheakoski** [comment]
I recently learned on http://php.net/manual/en/functions.anonymous.php

> As of PHP 5.4.0, when declared in the context of a class, the current class is automatically bound to it, making $this available inside of the function's scope.

which means that the queue processor itself can utilize DI and more complex code can be broken down into class methods on the processor instance instead of putting all of the code inside of the closure.

```php
class MyConsumer implements PsrProcessor
{
    protected $em;
    protected $jobRunner;

    public function __construct(
        EntityManagerInterface $em,
        JobRunner $jobRunner
    ) {
        $this->em = $em;
        $this->jobRunner = $jobRunner;
    }

    public function process(PsrMessage $message, PsrContext $session)
    {
        $id = $message->getBody();
        if (!preg_match('/\A\d+\z/', $id)) {
            return self::REJECT;
        }
        $result = $this->jobRunner->runUnique("job-$id", "myjob", function () use ($id) {
            return $this->doTheWork($id);
        });
        return $result ? self::ACK : self::REJECT;
    }

    protected function doTheWork(string $id): bool
    {
        $repo = $this->em->getRepository(MyEntity::class);
        /** @var MyEntity|null $entity */
        $entity = $repo->find($id);
        if (!$entity) {
            return false;
        }
        // ... more work ...
        return true;
    }
}
```

**Comment by makasim** [comment]
Could you open a PR?

**Comment by msheakoski** [comment]
The above example works without any changes. I was just sharing a workaround that I found for the problem that I faced initially once I learned that `$this` is accessible as the queue processor instance from inside the closure.

I think that `callable` is still better than `Closure` because it gives the end user more flexibility for how they want to design their queue processor, such as moving the logic to an external class or domain object. I need to look through the enqueue code to see if anything specifically depends on closures then I will submit a PR if the change is okay to do.
